### PR TITLE
sha2-mb

### PR DESCRIPTION
**Pull request by polubelova**
this PR introduces Multi-Buffer SHA2 implementation (Scalar, AVX, AVX2)

### CODE REVIEW

**Comment by prosecco** [comment]
Let's rename the folder to code/sha2-mb

File: code/sha2-vec/Hacl.Impl.SHA2.Generic.fst
Comment (msprotz) [inline code/sha2-vec/Hacl.Impl.SHA2.Generic.fst:32]: any way we can share this with the old-sha2 and put it into a static header?

File: code/sha2-vec/Hacl.Spec.SHA2.EquivScalar.fst
Comment (msprotz) [inline code/sha2-vec/Hacl.Spec.SHA2.EquivScalar.fst:608]: aren't these somewhere already?

File: code/sha2-vec/Hacl.Spec.SHA2.fst
Comment (msprotz) [inline code/sha2-vec/Hacl.Spec.SHA2.fst:11]: this seems to be reusing a lot of definitions already present elsewhere

**Comment by Kachoc** [comment]
Very neat code indeed! I agree with @protz  about trying not to duplicate code with the old SHA.
The most important remarks have already been made but I have a few (minor) comments.

This is more for information, because so far I don't see how I could ask for @polubelova to modify her code to account for my future PR, but I noticed there will likely be collisions with the work started by @R1kM on incremental blake and which I took on. So let's say we'ill probably have to coordinate a bit to minimize the pain of merging those branches (or put differently, once this branch gets merged into master, I'll probably have a few questions to prepare my PR - if as much code as possible is factorized, as suggested @protz , it will probably help me a lot, though).
For instance, there is [Hacl.Spec.SHA2.Equiv.fst](https://github.com/project-everest/hacl-star/blob/e260a1dbfe36cbb5ef95f2672a8f91d123bab17e/code/sha2-vec/Hacl.Spec.SHA2.Equiv.fst) which will (I think) overlap with [Spec.Hash.Incremental.Lemmas.fst](https://github.com/project-everest/hacl-star/blob/son_blake/specs/lemmas/Spec.Hash.Incremental.Lemmas.fst). Some other definitions like [update_last](https://github.com/project-everest/hacl-star/blob/e260a1dbfe36cbb5ef95f2672a8f91d123bab17e/code/sha2-vec/Hacl.Spec.SHA2.fst#250) will probably overlap too. More generally, the last block size is computed differently in the incremental blake branch, to make sure there is always some data to give to `update_last`. Still, I think the overlapping is definitely bearable on my side.

Another small remark which is not specifically about this branch but which I drop here so that it may be heard and cause a change in programming practices: I think it is not a good idea to declare long sequences definitions like the ones from [Spec.SHA2.Constants.fst](https://github.com/project-everest/hacl-star/blob/karthik_sha2_vec/specs/Spec.SHA2.Constants.fst) as `inline_for_extraction`. The reason is that at extraction time, many spec functions which more or less directly reference those definitions and are present in assertions in the code get expanded, causing significant blowups for the normalizer (I encountered the issue for blake). I'd rather advocate for using `normalize` (or else) at the exact place where they actually need to be expanded (i.e.: where we allocate the arrays in the code).

**Comment by msprotz** [comment]
Thanks @Kachoc -- regarding the list issue, is there any chance you could push a commit on this PR to establish a style, with comments on how to do it in a more optimal way in the future? Even just one instance might be enough to make sure future code is written in the way your propose. (Or just copy-paste a diff here or on slack.) Thanks for looking at this!

**Comment by Kachoc** [comment]
@protz regarding the list issue, for Blake I remove the `inline_for_extraction` from the `list_sigma` definition in `Spec.Blake2.fst`:
```
noextract
let list_sigma: list_sigma_t =
  [@inline_let]
  let l : list sigma_elt_t = [
    size  0; size  1; size  2; size  3; size  4; size  5; size  6; size  7;
    (* ... *)
    ] in
  assert_norm(List.Tot.length l == 160);
  l
```

And I updated the buffer definition in `Hacl.Impl.Blake2.Constants.fst` to look like that:

```
let sigmaTable : x:glbuffer Spec.sigma_elt_t 160ul{witnessed x Spec.sigmaTable /\ recallable x} =
  (**) norm_spec [delta_only [`%Spec.list_sigma]] Spec.list_sigma;
  createL_global (norm [delta_only [`%Spec.list_sigma]] Spec.list_sigma)
```

I think the `delta_only` is unnecessarily precise (on such a definition we could probably just use `normalize`)

**Comment by Kachoc** [comment]
Update on the list issue: inserting manual calls to `norm` and its derivatives actually doesn't work, because they overwrite the normalization parameters used by Kremlin at extraction time and thus prevent proper inlining of the definitions tagged `inline_for_extraction`. The correct method is to use post-processing (apply the normalization to update the definition body right after it has been defined):
```
noextract
let pp_sigmaTable () : Tactics.Tac unit =
  Tactics.norm [delta_only [`%Spec.list_sigma]]; Tactics.trefl ()

[@(Tactics.postprocess_with pp_sigmaTable)]
let sigmaTable : x:glbuffer Spec.sigma_elt_t 160ul{witnessed x Spec.sigmaTable /\ recallable x} =
  createL_global Spec.list_sigma
```

**Comment by karthikbhargavan** [comment]
Let us merge this. We can then address the concerns in another PR that integrates sha2-vec with the Agile API.

File: dist/gcc-compatible/Hacl_SHA2_Vec256.h
Comment (beurdouche) [inline dist/gcc-compatible/Hacl_SHA2_Vec256.h:52]: @polubelova Can you do something about these?

File: dist/gcc-compatible/libintvector.h
Comment (beurdouche) [inline dist/gcc-compatible/libintvector.h:313]: Is that really clang specific?

File: lib/Lib.IntVector.fst
Comment (beurdouche) [inline lib/Lib.IntVector.fst:566]: Why are those admitted?

File: specs/Spec.SHA2.fst
Comment (beurdouche) [inline specs/Spec.SHA2.fst:179]: Argh, is this split really necessary? If it really is, no worries... :)

**Review by beurdouche — APPROVED** [review]

File: lib/Lib.MultiBuffer.fst
Comment (msprotz) [inline lib/Lib.MultiBuffer.fst:25]: FYI there's a library to do this automatically rather than write all pairwise combinations by hand: https://github.com/FStarLang/FStar/blob/master/ulib/FStar.BigOps.fsti -- see https://github.com/FStarLang/FStar/blob/master/ulib/LowStar.Monotonic.Buffer.fsti#L1004 for an example of usage

File: lib/Lib.Sequence.fst
Comment (msprotz) [inline lib/Lib.Sequence.fst:414]: same remark here, you can meta-program those:
https://github.com/FStarLang/FStar/blob/master/ulib/FStar.Seq.Properties.fsti#L666
is something I wrote in support of this style, although the line number may lead you to believe this technique is cursed

**Review by msprotz — APPROVED** [review]
This looks good to me, minor the two little remarks I had -- happy to help with those if you're interested in having a solution that works once and for all, rather than having to write disjoint32 very soon ;-).

Thanks for moving the constants into a static header, this will probably do a lot for performance.

I have two requests.
- Can you open a bug and list precisely in a bullet-point style what remains to be done to get rid of the old implementation? That way we'll know where to start once someone has time to get rid of the old stuff. As you saw trying to merge this again, it's hard to get back into some old piece of code you haven't touched in a while. So, a description of "next tasks" would be good. I'm happy to add my own list of tasks to the bug once you're done with yours.
- I would like to see even a brief paragraph describing this new API. The argument names to your function are two-letter long (e.g. `b1`) and I actually doubt that anyone will figure out how this API works on their own. It'd be a shame if no one used your great work because they can't figure out how to use the API! So can you please add a word to the documentation? https://hacl-star.github.io/HaclHash.html -- it should be as simple as modifying this: https://github.com/project-everest/hacl-star/blob/master/doc/HaclHash.rst

Happy to help with any of these tasks, please let me know. Congratulations again on a really nice piece of work!

**Review by karthikbhargavan — APPROVED** [review]

**Comment by karthikbhargavan** [comment]
Marina, let's address Jonathan's 2 requests (open a bug, and document the API) together. I can help next week.

File: Makefile.common
Comment (polubelova) [reply Makefile.common]: you are right. i've removed Hacl.Hash.* from sha2vec_bundle

File: dist/gcc-compatible/libintvector.h
Comment (polubelova) [reply dist/gcc-compatible/libintvector.h]: removed from this PR

File: lib/Lib.Sequence.fst
Comment (polubelova) [reply lib/Lib.Sequence.fst:400]: thanks for the link! it works like a charm.

**Comment by polubelova** [comment]
thanks a lot for all the comments and suggestions!

@karthikbhargavan was working on the Low* implementation of sha2-mb, maybe he can address some of the above comments.

Karthik, sure, we can discuss it next week.

**Comment by prosecco** [comment]
shall we merge this?

**Comment by msprotz** [comment]
If you document the API and list the remaining items in a bug, you have my blessing! :-)

**Comment by karthikbhargavan** [comment]
Protz want a documented API which should not be hard to add. Bitrot is a bigger concern.

**Comment by msprotz** [comment]
I'm strongly in favor of merging this, but please write a minimum amount of documentation and list the remaining items, otherwise the knowledge will be lost and we'll totally struggle to remember what needs to be done to clean up. There will be quite a bit of duplication after this merge, so I want to make sure the issues are properly documented.

I can help with listing the remaining issues (brief call together?) and the documentation shouldn't take more than 10 minutes. Not something exhaustive, just a new page in the doc that explains what multi-buffer is, where to find it, and how to use it.

Thanks.

**Pull request merged**
